pennylane/operation.py

```diff
@@ -1130,17 +1135,29 @@ class Operation(Operator):
 
     Operations define some additional properties, such as differentiation
 
-    The following two class attributes are optional, but in most cases
-    should be clearly defined to avoid unexpected behavior during
-    differentiation.
-
-    * :attr:`~.Operation.grad_method`
-    * :attr:`~.Operation.grad_recipe`
-
-    Finally, there are some additional optional class attributes
-    that may be set, and used by certain quantum optimizers:
-
-    * :attr:`~.Operation.generator`
+    The first optional class attribute is :attr:`~.Operation.grad_recipe`.
+    It allows the definition of a hard-coded differentiation rule, which
+    typically will encode a parameter-shift rule. Note, however, that
+    operations will use ``grad_recipe`` preferably, so that the rule
+    can not be modified dynamically at runtime.
+
+    The second optional attribute is :attr:`~.Operation.parameter_frequencies`.
+    It allows to dynamically create differentiation rules for the operation,
+    respecting runtime settings for shift rules.
+    In addition, certain quantum optimizers can make use of ``parameter_frequencies``.
+
+    The third optional information is :attr:`~.Operation.generator`.
+    On one hand it will be used by quantum-aware optimizers and for Hadamard
+    test-based computation of differentiated quantities like the
+    :func:`~.transforms.metric_tensor`.
+    On the other hand, the eigenvalues of the ``generator`` fully determine the
+    ``parameter_frequencies`` and they will be obtained from it as a fallback
+    method.
+
+    Finally, the optional attribute :attr:`~.Operation.grad_method` can be used
+    to specify the differentiation behaviour. By default, it will be based
+    on whether the operation has parameters and which of the three above
+    attributes it provides.
```

Comment:
@dwierichs this is very nice information, _but_ I would argue it doesn't fit so well here; it is very dependent on the logic of `qml.gradients.param_shift`, and should probably go there (as suggested above).

Instead, I suggest reverting this to something similar to before, but with links to get more information:

```rest
The following two class attributes are optional, but in most cases
should be clearly defined to avoid unexpected behavior during
differentiation.

* :attr:`~.Operation.grad_recipe`
* [fill in as needed, or delete if this list no longer makes sense!]

Finally, there are some additional optional class attributes
that may be set, and used by certain transforms, quantum optimizers,
and gradient methods:

* :attr:`~.Operation.generator`
* :attr:`~.Operation.parameter_frequencies`

For details on how these are used during differentiation and other transforms,
please see the documentation for :class:`~.gradients.param_shift`,
:class:`~.metric_tensor`, :func:`~.reconstruct`
```

Note: the suggestion above is very rough! modify/change as needed :)

Reply:
I'm not sure how to write this most concisely while capturing our intention here:
While `grad_recipe` is the first-preference property used for differentiation, we decided to not recommend
manual addition of recipes because they are static and will (in the first iteration) not work with the `set_shifts` context etc.
Therefore, it feels wrong to recommend users to define `grad_recipe`.
I'll try to find a compromise.

Reply:
Yeah I agree with your assessment, I also wasn't entirely sure 🤔 

My main thinking here is that adding documentation about how parameter-shift and other functions use this info felt a bit too far removed and hard to maintain, but I don't have much better suggestions.

Feel free to make a judgement call here

Reply:
I think it is quite concise now, feel free to modify :)